### PR TITLE
HPC-7811 set up test environment

### PR DESCRIPTION
**Pull request by manelcecs**
This PR aims to add Jest test environment for HPC-API
Also provide a set of initial test for GraphQL

### CODE REVIEW

**Review by unocha-hpc — APPROVED** [review]
Checks have passed and this pull request is ready for manual review

**Review by unocha-hpc — APPROVED** [review]
Checks have passed and this pull request is ready for manual review

**Review by unocha-hpc — APPROVED** [review]
Checks have passed and this pull request is ready for manual review

**Review by unocha-hpc — APPROVED** [review]
Checks have passed and this pull request is ready for manual review

**Review by unocha-hpc — APPROVED** [review]
Checks have passed and this pull request is ready for manual review

File: bin/test.sh [hunk 1]
```diff
@@ -0,0 +1,87 @@
+root=$(pwd)
+
+#Global variables
+USAGE='Usage: test.sh [options] [-- [options]].\n Options:\n   -oc, --only-containers: only start docker containers\n    -sc, --stop-containers: stop docker containers\n    -k, --keep: keep jest runing after the completion of the tests suites\n    -c: run tests with coverage\n   -h, --help: show this help message\n    --: pass extra options'
+KEEP=0
+FORCE_STOP_JEST='--forceExit'
+ONLY_CONTAINERS=0
+STOP_CONTAINERSq=0
```
Comment (Pl217) [inline bin/test.sh]: Unused and has weird "q" at the end
```suggestion
```

File: bin/test.sh [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    echo 'Stopping docker containers'
+    moveToTestDir
+    docker-compose down
+    exit 0
+fi
+
+## ONLY_CONTAINERS must be 1 and STOP must be 0
+if [ $ONLY_CONTAINERS -eq 1 ] && [ $STOP -eq 1 ]; then
+    echo 'Invalid options - when using option -oc, option -ns must be used as well'
+    echo "$usage"
```
Comment (Pl217) [inline bin/test.sh]: This variable is named with all uppercases:
```suggestion
    echo "$USAGE"
```

File: bin/test.sh [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+if [ $STOP_CONTAINERS -eq 1 ]; then
+    echo 'Stopping docker containers'
+    moveToTestDir
+    docker-compose down
+    exit 0
+fi
+
+## ONLY_CONTAINERS must be 1 and STOP must be 0
+if [ $ONLY_CONTAINERS -eq 1 ] && [ $STOP -eq 1 ]; then
+    echo 'Invalid options - when using option -oc, option -ns must be used as well'
```
Comment (Pl217) [inline bin/test.sh]: Variable `$STOP` doesn't exist. Also, option `ns` that you mention in error message doesn't exist too.

File: tests/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        soft: 65000
+        hard: 65000
+  hid:
+    image: public.ecr.aws/unocha/nodejs:18-alpine
+    volumes:
+      - ../tools/hid_api_mock:/srv/www
+    environment:
+      PORT: 80
+    hostname: hidapi
+    container_name: test_hidservice_test_api
```
Comment (Pl217) [inline tests/docker-compose.yml]: The HID API mock service doesn't even exist in this repo

File: tests/docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    image: solr:8.11.2-slim
+    environment:
+      DNSDOCK_NAME: solr
+      DNSDOCK_IMAGE: hpc
+      SOLR_CORE: fts ## take it as env var SOLR_CORE: ${SOLR_CORE}
+    container_name: hpcservice_solr_1_test_api
+    volumes:
+      - '../config/services/solr:/opt/hpc-solr:rw'
+      - '../data/test/hpc/solr_logs:/var/solr/logs:rw'
+      - '../data/test/hpc/solr_backups:/var/solr/data/backups:rw'
```
Comment (Pl217) [inline tests/docker-compose.yml]: We don't need to be concerned with Solr backups for test containers.

File: bin/test.sh [hunk 6]
```diff
@@ -0,0 +1,87 @@
+root=$(pwd)
+
+#Global variables
+USAGE='Usage: test.sh [options] [-- [options]].\n Options:\n   -oc, --only-containers: only start docker containers\n    -sc, --stop-containers: stop docker containers\n    -k, --keep: keep jest runing after the completion of the tests suites\n    -c: run tests with coverage\n   -h, --help: show this help message\n    --: pass extra options'
```
Comment (Pl217) [inline bin/test.sh]: We can have multi-line text in shell scripts.

File: bin/test.sh [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+                          done
+                          ;;
+    * )                   echo "$USAGE"
+                          exit 1
+  esac
+  shift
+done
+
+## STOP_CONTAINERS is a final option
+if [ $STOP_CONTAINERS -eq 1 ]; then
```
Comment (Pl217) [inline bin/test.sh]: I am getting an error here: `./bin/test.sh: line 49: [: : integer expression expected`

File: tests/resolvers/software-info.spec.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        query: 'query { softwareInfo { title status version } }',
+      });
+
+    expect(response).toBeDefined();
+    expect(response.errors).toBeUndefined();
+    expect(response.data).toBeDefined();
+    const data = response.data as any;
+    expect(data.softwareInfo).toBeDefined();
+    expect(data.softwareInfo.length).toBeGreaterThan(0);
+    const softwareInfo = data.softwareInfo[0];
```
Comment (Pl217) [inline tests/resolvers/software-info.spec.ts]: We can remove usage of `any` and use [optional chaining operator (`?.`)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining).
```suggestion
    const data = response.data;
    expect(data?.softwareInfo).toBeDefined();
    expect(data?.softwareInfo.length).toBeGreaterThan(0);
    const softwareInfo = data?.softwareInfo[0];
```

File: tests/resolvers/software-info.spec.ts
Comment (Pl217) [inline tests/resolvers/software-info.spec.ts:1]: The test cases in this file are pretty repetitive and you can extract the code into a common method.

File: tests/docker-compose.yml
Comment (Pl217) [inline tests/docker-compose.yml:1]: We need to update this file for Compose V2 - HPC-9088

File: tests/testContext.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+  transactions: Record<string, Knex.Transaction>;
+  apolloTestServer: ApolloServer;
+}
+
+export default class ContextProvider implements IContext {
+  private static _instance: ContextProvider;
+
+  models?: Database;
+  conn: Knex;
+  transactions: Record<string, Knex.Transaction>;
```
Comment (Pl217) [inline tests/testContext.ts]: The transactions aren't used anywhere, so we can just remove the property.
```suggestion
```

**Review by Pl217 — CHANGES_REQUESTED** [review]
I have rebased this PR and resolved issues I write about inline. All except the comment about undefined variable `$STOP` and option `-ns`, which don't exist, but are mentioned in `bin/test.sh`.

Also, I have made the following changes:
- Set up CI to run unit tests
- Changed Solr config to latest one
- Use newest DB migration file

**Review by unocha-hpc — APPROVED** [review]
Checks have passed and this pull request is ready for manual review

**Review by unocha-hpc — APPROVED** [review]
Checks have passed and this pull request is ready for manual review

**Review by unocha-hpc — APPROVED** [review]
Checks have passed and this pull request is ready for manual review

**Review by Pl217 — APPROVED** [review]
Me and @manelcecs have agreed to remove Solr from test containers.

**Pull request merged**